### DISCUSSION TITLE
Chained Tasks Not Executed In-order

### DISCUSSION THREAD

**Discussion by arshad171** [celery/celery#7114] [Q&A]
For some of my TensorFlow pipelines, I need the task chains to be executed **in the order they were received**, i.e. execute the entire chain before beginning with the next chain scheduled.

I have abstracted the problem with the example below,

## Sample Tasks

```
@app.task
def task1(id):
    print(f'{id=} | task 1')
    time.sleep(5)
    return id

@app.task
def task2(id):
    print(f'{id=} | task 2')
    time.sleep(5)
    return id
```

## Chaining Tasks

Creating two chains, with different ids, `id=1` and `id=2`.

```
res1 = chain(
    task1.s(1),
    task2.s()
).apply_async()

res2 = chain(
    task1.s(2),
    task2.s()
).apply_async()
```

### The result I expect: The first chain (with id=1) to be processed before the chain with id=2 is picked up.

```
[2021-11-22 20:42:43,653: WARNING/ForkPoolWorker-1] id=1 | task 1
[2021-11-22 20:42:48,663: WARNING/ForkPoolWorker-1] id=1 | task 2
[2021-11-22 20:42:53,677: WARNING/ForkPoolWorker-1] id=2 | task 1
[2021-11-22 20:42:58,685: WARNING/ForkPoolWorker-1] id=2 | task 2
```

### What I observe: task1 from both the chains are run first, and then the task2 is picked up.

```
[2021-11-22 20:42:43,653: WARNING/ForkPoolWorker-1] id=1 | task 1
[2021-11-22 20:42:48,663: WARNING/ForkPoolWorker-1] id=2 | task 1
[2021-11-22 20:42:53,677: WARNING/ForkPoolWorker-1] id=1 | task 2
[2021-11-22 20:42:58,685: WARNING/ForkPoolWorker-1] id=2 | task 2
```

### I have also tried passing `--concurrency=1` and `--prefetch-multiplier=1` when starting the worker, but it doesn't seem to help, same result as above.

Is this the expected behaviour?

Please suggest any ways to achieve what I'm in need of..

## Setup

- Python: 3.8
- Celery: 5.2.1
- Backend: RabbitMQ
- OS: Linux (Manjaro)

Thanks

**Comment by auvipy** [comment]
could you please try the latest pypi release in the mean time?

**Reply by arshad171** [reply]
Sorry the description says celery v5.1.2, but I meant **5.2.1**, I'll update the description.
As listed under the "Setup", this behaviour is seen in the latest version of celery, v5.2.1 (latest pypi release).

**Reply by auvipy** [reply]
did you also checked duplicate issues using canvas tag/label?

**Reply by arshad171** [reply]
yes, I did a very extensive search before raising this discussion (and converting it to an issue), I could not find any similar discussion/issue.

**Comment by januwA** [comment]
I think the order of execution is correct

**Reply by auvipy** [reply]
i think it's answered here https://github.com/celery/celery/issues/7117#issuecomment-1023054262
